src/start.sh

```diff
@@ -75,33 +97,63 @@ start() {
     # Start tailing the FTL log from the most recent "FTL Started" message
     # Get the line number
     startFrom=$(grep -n '########## FTL started' /var/log/pihole/FTL.log | tail -1 | cut -d: -f1)
-    # Start the tail from the line number
-    tail -f -n +${startFrom} /var/log/pihole/FTL.log &
+    # Start the tail from the line number and background it
+    tail --follow=name --retry -n +${startFrom} /var/log/pihole/FTL.log &
   else
     echo "  [i] FTL log output is disabled. Remove the Environment variable TAIL_FTL_LOG, or set it to 1 to enable FTL log output."
   fi
 
-  # https://stackoverflow.com/a/49511035
-  wait $!
+  # Wait for the ftl process to finish and handle its return
+  wait $ftl_pid
+  stop $?
 }
 
 stop() {
-  # Ensure pihole-FTL shuts down cleanly on SIGTERM/SIGINT
-  ftl_pid=$(pgrep pihole-FTL)
-  killall --signal 15 pihole-FTL
+  local FTL_EXIT_CODE=$1
 
-  # Wait for pihole-FTL to exit
-  while test -d /proc/"${ftl_pid}"; do
-    sleep 0.5
-  done
+  # Only attempt to close pihole-FTL if it is running, it may already have crashed
+  if pgrep pihole-FTL >/dev/null; then
+    echo ""
+    echo "  [i] Container stop requested..."
+    echo "  [i] pihole-FTL is running - Attempting to shut it down cleanly"
+    echo ""
+    # Ensure pihole-FTL shuts down cleanly on SIGTERM/SIGINT
+    ftl_pid=$(pgrep pihole-FTL)
+
+    killall --signal 15 pihole-FTL    
+
+    # Wait for pihole-FTL to exit
+    while test -d /proc/"${ftl_pid}"; do
+      sleep 0.5
+    done
+
+    # Assume FTL exited cleanly at this point
+    FTL_EXIT_CODE=0
```

Comment:
Assuming might be right, checking is better.

```suggestion
     tail --pid=$ftl_pid -f /dev/null 
    FTL_EXIT_CODE=$?
```

Not using `wait` here because of it will complain about pid not child of this shell.
https://unix.stackexchange.com/a/427133

Reply:
👀👀 nice. Will play with this later.

Reply:
Mind that the return code of `tail` will be ... well ... the return code of `tail` here. Which is `0`  once the process exited - regardless *how* it exited, see [here](https://github.com/coreutils/coreutils/blob/8083944484f2cdf6c9b737642567bcdb54db784d/src/tail.c#L1656).

The Bash built-in `wait`, however, returns the exit status of the *last* command/process waited for.

Reply:
Good point, but

```
2024-10-09 10:29:59.883 UTC [164M] INFO: Stored 0 API sessions in the database
2024-10-09 10:30:00.085 UTC [164M] INFO: ########## FTL terminated after 4s 184ms  (code 0)! ##########
/usr/bin/start.sh: line 130: wait: pid 164 is not a child of this shell

  [i] pihole-FTL exited with status 127

  [i] Container will now stop or restart depending on your restart policy
```

Reply:
Sure, I don't think there is a way to (easily) achieve what we want here - it is really an issue of two independent shell sessions. There is nothing in the Linux kernel I'd be aware of that allows obtaining the exit code of a process that has already exited and which was not a child of our process. There are low(er)-level standard library C functions available that allow doing such things and I was almost ready to suggest let's add an undocumented flag to `pihole-FTL`, but then I found deep in their description the part

> [...] wait for state changes in a child of the calling process [...]

which is the exact limitation `wait [pid]` has (which makes me assuming - without checking - that `bash` implements this feature using exactly these stdlib functions. The code even specifies that if you nonetheless try to use these functions for the described purpose, the functions will immediately return with `ECHILD` indicating this parent/child requirement is violated here.

---

I guess the only way to overcome this would be "transfering" the knowledge from the one terminal session that was/is still running `start.sh` and then storing the exit code of FTL somewhere on **disk** (like in `/tmp/ftl_exit_code`) from where the `stop()` function can retrieve it once it is there. This is obviously a rather dirty solution but the only one that comes to mind now.

**edit** Well, another solution would be sourcing the value from the `FTL.log` file. On exit/crash/whatever, we know that the very last line will be
```
2024-10-09 10:30:00.085 UTC [164M] INFO: ########## FTL terminated after 4s 184ms  (code 0)! ##########
```
maybe like
``` bash
# Extract the code from the last line of FTL's log file
code=$(tail -n 1 "$log_file" | grep -oP '\(code \K[0-9]+')
```

Reply:
> This is obviously a rather dirty solution but the only one that comes to mind now.

This was what I was originally doing, but also agree it is a bit dirty

Reply:
Latest commit should fix all of this